src/concurrency/genmc/intercept.rs

```diff
@@ -9,64 +10,56 @@ use crate::{
 
 // Handling of code intercepted by Miri in GenMC mode, such as assume statement or `std::sync::Mutex`.
 
-impl<'tcx> EvalContextExt<'tcx> for crate::MiriInterpCx<'tcx> {}
-pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
-    /// Given a `ty::Instance<'tcx>`, do any required special handling.
-    /// Returns true if this `instance` should be skipped (i.e., no MIR should be executed for it).
-    fn check_genmc_intercept_function(
-        &mut self,
-        instance: rustc_middle::ty::Instance<'tcx>,
-        args: &[rustc_const_eval::interpret::FnArg<'tcx, crate::Provenance>],
-        dest: &crate::PlaceTy<'tcx>,
-    ) -> InterpResult<'tcx, bool> {
-        let this = self.eval_context_mut();
-        let genmc_ctx = this
-            .machine
-            .data_race
-            .as_genmc_ref()
-            .expect("This function should only be called in GenMC mode.");
-
-        let get_mutex_call_infos = || {
-            assert_eq!(args.len(), 1);
-            let arg = this.copy_fn_arg(&args[0]);
-            let addr = this.read_target_usize(&arg)?;
-            // FIXME(genmc): assert that we have at least 1 byte.
-            // FIXME(genmc): maybe use actual size of mutex here?.
-
-            let thread_infos = genmc_ctx.exec_state.thread_id_manager.borrow();
-            let curr_thread = this.machine.threads.active_thread();
-            let genmc_curr_thread = thread_infos.get_genmc_tid(curr_thread);
-            interp_ok((genmc_curr_thread, addr, 1))
-        };
+#[derive(Clone, Copy)]
+struct MutexMethodArgs {
+    address: u64,
+    size: u64,
+}
 
-        use rustc_span::sym;
-        interp_ok(if this.tcx.is_diagnostic_item(sym::sys_mutex_lock, instance.def_id()) {
-            debug!("GenMC: handling Mutex::lock()");
-            let (genmc_curr_thread, addr, size) = get_mutex_call_infos()?;
+impl<'tcx> EvalContextExtPriv<'tcx> for crate::MiriInterpCx<'tcx> {}
+trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+    fn parse_mutex_method_args(
+        &self,
+        args: &[rustc_const_eval::interpret::FnArg<'tcx, crate::Provenance>],
+    ) -> InterpResult<'tcx, MutexMethodArgs> {
+        assert_eq!(args.len(), 1, "Mutex lock/unlock/try_lock should take exactly 1 argument.");
+        let this = self.eval_context_ref();
+        let arg = this.copy_fn_arg(&args[0]);
+        // FIXME(genmc): use actual size of the pointee of `arg`.
+        let size = 1;
+        // GenMC does not support large accesses, we limit the size to the maximum access size.
```

Comment:
Something is very strange if the mutex is too big for this.

Papering over that by clamping the size doesn't sound good. Let's just have an assertion in `handle_mutex_lock` (like we already have for some other handlers I think?)